Add signposts and profiling support to JetStream2

xw opened the pull request:
#### c6a3a8988a7f1411cab94e77b52f84363d9486c0
<pre>
Add signposts and profiling support to JetStream2
<a href="https://bugs.webkit.org/show_bug.cgi?id=247907">https://bugs.webkit.org/show_bug.cgi?id=247907</a>
rdar://102328951

Reviewed by Dewei Zhu.

Two things are needed to add profiling support to JetStream:
  - Support for subtests, since by default trying to profile JetStream may sometimes result in the profile being cut off before the test finishes due to file size issues.
  - Signposting to help label generated profiles

In this patch:
  - Added two new fields to plan files in order to allow them to support subtests
                - `subtest_url_format`: a string template for formatting subtest URL parameters
                - `subtests`: a dictionary of suites and tests in the following form:
                         {&quot;suite1&quot;: [&quot;test1, test2&quot;]}
                  If a test does not have a suite associated with it, the key is &quot;&quot;.
  - Added a new argument to run-benchmarks, `--list-subtests`, that prints out available subtests from a specified plan.
        - Modified and added helper functions to BenchmarkRunner to support listing and validating subtests.
  - Added a new argument to run-benchmarks, `--subtests`, that limits a test with subtest support to only run those subtests.
  - Added a new signposting patch to help annotate generated profiles.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner.__init__):
(BenchmarkRunner):
(BenchmarkRunner._find_plan_file):
(BenchmarkRunner._load_plan_data):
(BenchmarkRunner.format_subtests):
(BenchmarkRunner.available_subtests):
(BenchmarkRunner.validate_subtests):
* Tools/Scripts/webkitpy/benchmark_runner/data/patches/signposts/JetStream2.patch: Added.
* Tools/Scripts/webkitpy/benchmark_runner/data/plans/jetstream2.1.plan:
* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(config_argument_parser):
(run_benchmark_plan):
(list_subtests):
(start):
* Tools/Scripts/webkitpy/benchmark_runner/webdriver_benchmark_runner.py:
(WebDriverBenchmarkRunner._construct_subtest_url):
(WebDriverBenchmarkRunner._run_one_test):
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner.__init__):
(WebServerBenchmarkRunner):

Canonical link: <a href="https://commits.webkit.org/257642@main">https://commits.webkit.org/257642@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d02cd4daf5b4991494972f6bc51a578b47d0e8d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99564 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32657 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108926 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/169163 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103557 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9325 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86044 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/106843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105324 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/92038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/103074 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/2580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2512 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5258 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/8660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/42914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4372 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xw commented on Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
```diff
@@ -85,6 +92,36 @@ def available_plans():
+        valid_subtests = self._plan['subtests']
+        subtests_to_run = []
+        for subtest in subtests:
+            if '/' in subtest:
```
Although this doesn't affect this patch at all (since JetStream subtests are not organized by suites as far as I can tell), I'm open to feedback on what the best way to require users to format subtest names is.

Currently it's case sensitive `SuiteName/TestName` (or just `TestName` if the plan does not specify suites, like JetStream) - but perhaps `SuiteName:TestName` would be better? Or alternatively, should this be made more robust so that we should be able to infer a `SuiteName` when one isn't passed in?

Again, this consideration doesn't affect JetStream at all (it will be a more pertinent for when we add subtest support for MotionMark in run-benchmarks), so maybe there is also a case for postponing this particular discussion and making a note to iterate on it in the future.

dewei-zhu replied:
If we support suite name, will we support using a suite name as a short hand for all subtests under certain suite?
I have no strong preference on the syntax itself.

xw replied:
I'm conflicted on that. On one hand it'd be nice for QoL, but on the other if the user doesn't explicitly specify a suite name we should default to it being a test name (assuming we decide that it'd be best to infer a suite name when one isn't passed in). So on the off chance that a suite and a test both have the same name, it'll default to just running the one subtest instead of everything from the suite. Unless maybe it'd be a better to reverse this logic, and check suites first before individual subtests?

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
```diff
@@ -85,6 +92,36 @@ def available_plans():
+        return BenchmarkRunner.format_subtests(plan['subtests'])
+
+    def validate_subtests(self, subtests):
+        valid_subtests = self._plan['subtests']
```
This may throw an exception when plan file does not include subtest, maybe use `self._plan.get('subtests', [])`?

xw replied:
If the plan file does not include subtests and we pass in stuff through the `--subtests` flag, we want an exception to be thrown. If we don't pass in `--subtests`, we shouldn't hit this function in the first place (which happens to be currently bugged apparently, fix incoming shortly)

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
```diff
@@ -53,12 +53,13 @@ def config_argument_parser():
     parser.add_argument('--show-iteration-values', dest='show_iteration_values', action='store_true', help="Show the measured value for each iteration in addition to averages.")
     parser.add_argument('--generate-pgo-profiles', dest="generate_pgo_profiles", action='store_true', help="Collect LLVM profiles for PGO, and copy them to the diagnostics directory.")
     parser.add_argument('--profile', action='store_true', help="Collect profiling traces, and copy them to the diagnostic directory.")
-
```
Let's keep this blank line since below specifies a mutually exclusive group.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/data/plans/jetstream2.1.plan:
```diff
@@ -516,5 +517,76 @@
+    },
+    "subtest_url_format": "test=${TEST}",
+    "subtests": [
+        { "suite": "", "test": "Air" },
```
If we think suite will be a tree structure, why don't we use dictionary like 
`{"": ["Air", "Basic", ...]}`.
For test without concept of suite like `JetStream`, it seems cleaner to be
`"subtests": ["Air", "Basic"]`.
We can tell if the test have concept of suite from the type of the value under key `subtests`.

xw replied:
Part of the reason is to keep it logically consistent with the templating structure of `subtest_url_format` - the two valid arguments there should be `${TEST}` and `${SUITE}`. This is more relevant for something like MotionMark, where we'd have a template string like `&suite-name=${SUITE}&test-name=${TEST}` - so (to me, at least) it makes sense to have each subtest be its own little object with `suite` and `test` fields.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/data/plans/jetstream2.1.plan:
```diff
@@ -516,5 +517,78 @@
+            "lebab-wtb",
+            "prepack-wtb",
+            "uglify-js-wtb"
+        ], "test": ["test1", "test2"]
```
I guess this is added for debugging, we may want to remove this line.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
```diff
@@ -85,6 +93,52 @@ def available_plans():
+                    if subtest_name in tests:
+                        subtests_to_run[suite].append(subtest_name)
+                        did_append_subtest = True
+                        break
```
Does this mean we will use the first match if there are multiple tests with same name but under different suites?

xw replied:
Yep - first match alphabetically ordered by suite (or no suite, if one exists)

dewei-zhu reviewed (approved):
r=me with comment.

xw committed:
Add signposts and profiling support to JetStream2
https://bugs.webkit.org/show_bug.cgi?id=247907
rdar://102328951

Reviewed by Dewei Zhu.

Two things are needed to add profiling support to JetStream:
  - Support for subtests, since by default trying to profile JetStream may sometimes result in the profile being cut off before the test finishes due to file size issues.
  - Signposting to help label generated profiles

In this patch:
  - Added two new fields to plan files in order to allow them to support subtests
                - `subtest_url_format`: a string template for formatting subtest URL parameters
                - `subtests`: a dictionary of suites and tests in the following form:
                         {"suite1": ["test1, test2"]}
                  If a test does not have a suite associated with it, the key is "".
  - Added a new argument to run-benchmarks, `--list-subtests`, that prints out available subtests from a specified plan.
        - Modified and added helper functions to BenchmarkRunner to support listing and validating subtests.
  - Added a new argument to run-benchmarks, `--subtests`, that limits a test with subtest support to only run those subtests.
  - Added a new signposting patch to help annotate generated profiles.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner.__init__):
(BenchmarkRunner):
(BenchmarkRunner._find_plan_file):
(BenchmarkRunner._load_plan_data):
(BenchmarkRunner.format_subtests):
(BenchmarkRunner.available_subtests):
(BenchmarkRunner.validate_subtests):
* Tools/Scripts/webkitpy/benchmark_runner/data/patches/signposts/JetStream2.patch: Added.
* Tools/Scripts/webkitpy/benchmark_runner/data/plans/jetstream2.1.plan:
* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(config_argument_parser):
(run_benchmark_plan):
(list_subtests):
(start):
* Tools/Scripts/webkitpy/benchmark_runner/webdriver_benchmark_runner.py:
(WebDriverBenchmarkRunner._construct_subtest_url):
(WebDriverBenchmarkRunner._run_one_test):
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner.__init__):
(WebServerBenchmarkRunner):

Canonical link: https://commits.webkit.org/257642@main
